20251028-BRS-workflow-versioning.md

```diff
@@ -0,0 +1,724 @@
+
+*Specifics about how this can be implemented in different Dapr SDKs is covered in more detail below and is briefly 
+provided here to give a high-level overview of the approach.*
+
+Workflows and activities will be registered with the Dapr Workflow SDK as they are today.
+
+Whenever a request comes into the application from the Dapr runtime to run work on a specific workflow following a 
+`ScheduleNewWorkflowAsync` call, during the same following a `ContinueAsNew` call or even following a 
+`CallChildWorkflowAsync`, a request should come into the SDKs to run that named workflow.
+
+### Version isn't specified on orchestration request
+If there is no version specified on the event, the SDK should map the request based on the named base type to the
+latest version available in the application and run the workflow there. Every time a patch is evaluated, the "new code" 
+route should be taken, and the patch identifier persisted in a list of encountered patches.
+
+At each boundary invocation except workflow orchestration completion, the SDK should return the workflow status with
+the (potentially updated) version information defined below containing the name of the workflow type executed and the list 
+of named patches followed.
+
+If the patch does not exist in the version list, we should assume that we're being asked to evaluate the patch value
+for the first time. In this case, the following logic should be applied:
+
+A call to `IsPatched`:
+- Returns `true` if the workflow is not presently replaying
+- Returns `true` if the workflow is replaying and the patch name is present in the version information
+- Returns `false` if the workflow is replaying and the patch name is **not** present in the version information
+
+### Version is specified on orchestration request
+If the version is supplied, the workflow type should be parsed out and the SDK should ignore "latest" evaluations and use
+that type to run the workflow execution. The list of patches should be extracted from the version prototype. At each
+evaluation of the patch, several checks should happen:
+
+- If the patch exists in the version list, the "new code" route should be taken if it has a value of true. Otherwise, the
+"old code" route should be taken if it has a value of false.
+- Is the name of the patch present in an earlier `OchestratorStartedEvent` than the current message? If so, the workflow
+should fail with a `PENDING` status, indicating to the runtime that the code does not match the expected workflow state.
+- If the evaluation has gotten to this point, the workflow should fail with a `PENDING` status because again, the code
+being executed has run this workflow replay to completion before (as we saved the patch map information previously), so 
+if the patch name isn't present, the code has changed and does not match the workflow history.
+
+In case of a `PENDING` status, the runtime should wait to replay the workflow again until a dissemination event has 
```

Comment:
the `PENDING` status already exists. I think we want to have a specific status for this, like `VERSION_PENDING` or `PENDING_VERSION`.

Reply:
Id prefer `PENDING_VERSION` over version_pending. @WhitWaldo mind updating that verbiage one more time to: `PENDING_VERSION`? :)

Reply:
Given that this situation is only encountered when dealing with patching (not type versions), what about being even more specific with `PENDING_PATCHING`?